// @barchart/common-portfolio/engine/portfolio/operations/transactions/Income.js
const assert = require('@barchart/common-js/lang/assert'),
	attributes = require('@barchart/common-js/lang/attributes'),
	Decimal = require('@barchart/common-js/lang/Decimal'),
	is = require('@barchart/common-js/lang/is');

const Create = require('./Create'),
	Credit = require('./Credit'),
	TransactionType = require('./../../../../data/TransactionType');

module.exports = (() => {
	'use strict';

	/**
	 * Creates a {@link TransactionType.INCOME} transaction.
	 *
	 * @public
	 * @extends {Create}
	 * @param {Object} data
	 * @param {String} data.user - The user identifier (of the portfolio's owner).
	 * @param {String} data.portfolio - The identifier of the portfolio.
	 * @param {String} data.position - The identifier of the position.
	 * @param {Day} data.date - The day of the transaction.
	 * @param {Timestamp} data.timestamp - The moment of the transaction.
	 * @param {Decimal} data.income - The amount of income, in the instrument's currency.
	 */
	class Income extends Create {
		constructor(data) {
			super(data);

			assert.argumentIsValid(data.income, 'data.income', Decimal.getIsNotNegative, 'cannot be negative');
		}

		get type() {
			return TransactionType.INCOME;
		}

		_createTransaction(position, transactions, data) {
			return {
				income: {
					amount: data.income
				}
			};
		}

		_afterTransactionCreated(position, transactions, transaction) {
			const data = { };

			data.user = transaction.user;
			data.portfolio = transaction.portfolio;
			data.currency = position.instrument.currency;

			data.date = transaction.date;
			data.timestamp = transaction.timestamp;

			data.amount = transaction.income.amount;

			data.reference = {
				position: transaction.position,
				sequence: transaction.sequence
			};

			data.description = `Income on ${getDescriptionForInstrument(position.instrument)}`;

			if (is.object(transaction.legacy)) {
				const legacy = { };

				if (transaction.legacy.system) {
					legacy.system = transaction.legacy.system;
				}

				if (transaction.legacy.user) {
					legacy.user = transaction.legacy.user;
				}

				if (transaction.legacy.portfolio) {
					legacy.portfolio = transaction.legacy.portfolio;
				}

				data.legacy = legacy;
			}

			this._spawn(new Credit(data));
		}

		toString() {
			return '[Income]';
		}
	}

	function getDescriptionForInstrument(instrument) {
		if (attributes.has(instrument, 'symbol.display')) {
			return attributes.read(instrument, 'symbol.display');
		} else {
			return instrument.name || instrument.id;
		}
	}

	return Income;
})();
